src/app/util/attribute-storage.cpp

```diff
@@ -102,13 +102,16 @@ constexpr const chip::CommandId generatedCommands[] = GENERATED_COMMANDS;
 constexpr const EmberAfAttributeMetadata generatedAttributes[]      = GENERATED_ATTRIBUTES;
 constexpr const EmberAfCluster generatedClusters[]                  = GENERATED_CLUSTERS;
 constexpr const EmberAfEndpointType generatedEmberAfEndpointTypes[] = GENERATED_ENDPOINT_TYPES;
+constexpr const EmberAfDeviceType fixedDeviceTypeList[]             = FIXED_DEVICE_TYPES;
+
 // Not const, because these need to mutate.
 DataVersion fixedEndpointDataVersions[ZAP_FIXED_ENDPOINT_DATA_VERSION_COUNT];
 
 #if !defined(EMBER_SCRIPTED_TEST)
 #define endpointNumber(x) fixedEndpoints[x]
-#define endpointDeviceId(x) fixedDeviceIds[x]
-#define endpointDeviceVersion(x) fixedDeviceVersions[x]
+#define endpointDeviceTypeList(x)                                                                                                  \
+    Span<const EmberAfDeviceType>(&fixedDeviceTypeList[fixedDeviceTypeListOffsets[x]],                                             \
+                                  fixedDeviceTypeListLengths[fixedDeviceTypeListOffsets[x]])
```

Comment:
I would expect `fixedDeviceTypeListLengths[x]` for the length.....  Why are we indirecting via the offset?

Put another way, why do we have an entry per device type in the "lengths" list, not an entry per endpoint?

Reply:
Dang, good catch.

I did a bunch of manual tests after using the log-source app with all permutations of endpoints with device types vs. endpoints without just to validate this logic (of course, nothing that could be checked in unfortunately...)

Reply:
Isn't the problem that we can't even begin to do that because of ZAP limitations? The means to specify this in the ZAP file is itself broken and doesn't permit a list-like representation:

```
      "name": "MA-onofflight",
      "deviceTypeName": "MA-onofflight",
      "deviceTypeCode": 256,
      "deviceTypeProfileId": 259,
      "clusters": [
        {
          "name": "Identify",
...
...
```

Reply:
Yes, that is what makes testing this code fully hard for now...